File: apps/go/internal/handler/catalog.go
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/ryansuhartanto/koda-b8-backend/apps/go/internal/model"
	"github.com/ryansuhartanto/koda-b8-backend/apps/go/internal/repository"
)

// TODO: admin POST and PATCH for /categories and /brands go here
func Catalog(r *gin.Engine, pool *pgxpool.Pool) {
	r.GET("/categories", listCategories(pool))
	r.GET("/brands", listBrands(pool))
	r.GET("/shipping-methods", listShippingMethods(pool))
	r.GET("/payment-methods", listPaymentMethods(pool))
}

// @Summary  List categories and how many products each holds
// @Tags     catalog
// @Produce  json
// @Success  200 {array}  model.CategoriesSummary "OK"
// @Failure  500 {object} model.Problem  "Internal error"
// @Router   /categories [get]
func listCategories(pool *pgxpool.Pool) gin.HandlerFunc {
	return func(ctx *gin.Context) {
		categories, err := repository.Categories(ctx, pool)
		if err != nil {
			model.AbortProblem(ctx, http.StatusInternalServerError, err.Error())
			return
		}

		ctx.PureJSON(http.StatusOK, categories)
	}
}

// @Summary  List brands and how many products each holds
// @Tags     catalog
// @Produce  json
// @Success  200 {array}  model.BrandsSummary   "OK"
// @Failure  500 {object} model.Problem "Internal error"
// @Router   /brands [get]
func listBrands(pool *pgxpool.Pool) gin.HandlerFunc {
	return func(ctx *gin.Context) {
		brands, err := repository.Brands(ctx, pool)
		if err != nil {
			model.AbortProblem(ctx, http.StatusInternalServerError, err.Error())
			return
		}

		ctx.PureJSON(http.StatusOK, brands)
	}
}

// @Summary  List shipping methods and their cost
// @Tags     catalog
// @Produce  json
// @Success  200 {array}  model.ShippingMethod "OK"
// @Failure  500 {object} model.Problem        "Internal error"
// @Router   /shipping-methods [get]
func listShippingMethods(pool *pgxpool.Pool) gin.HandlerFunc {
	return func(ctx *gin.Context) {
		methods, err := repository.ShippingMethods(ctx, pool)
		if err != nil {
			model.AbortProblem(ctx, http.StatusInternalServerError, err.Error())
			return
		}

		ctx.PureJSON(http.StatusOK, methods)
	}
}

// @Summary  List the payment methods an order can be placed with
// @Tags     catalog
// @Produce  json
// @Success  200 {array}  model.PaymentMethod "OK"
// @Failure  500 {object} model.Problem       "Internal error"
// @Router   /payment-methods [get]
func listPaymentMethods(pool *pgxpool.Pool) gin.HandlerFunc {
	return func(ctx *gin.Context) {
		methods, err := repository.PaymentMethods(ctx, pool)
		if err != nil {
			model.AbortProblem(ctx, http.StatusInternalServerError, err.Error())
			return
		}

		ctx.PureJSON(http.StatusOK, methods)
	}
}

File: apps/js/src/routes/catalog.ts
import { Router } from "express";

import { fail } from "#/lib/problem";
import { wire } from "#/lib/wire";
import * as catalog from "#/service/catalog";

// TODO: admin POST and PATCH for /categories and /brands go here
export const router: Router = Router();

/**
 * @openapi
 * components:
 *   schemas:
 *     Category:
 *       type: object
 *       properties:
 *         id: { type: string }
 *         name: { type: string }
 *         icon: { type: string }
 *         img: { type: string }
 *         product_count: { type: integer }
 *       required: [id, name, product_count]
 *     Brand:
 *       type: object
 *       properties:
 *         id: { type: string }
 *         name: { type: string }
 *         product_count: { type: integer }
 *       required: [id, name, product_count]
 *     ShippingMethod:
 *       type: object
 *       properties:
 *         id: { type: string }
 *         name: { type: string }
 *         cost_idr: { type: integer }
 *       required: [cost_idr, id, name]
 *     PaymentMethod:
 *       type: object
 *       properties:
 *         id: { type: string }
 *         name: { type: string }
 *         metadata: { type: object }
 *       required: [id, metadata, name]
 *
 * /categories:
 *   get:
 *     summary: List categories
 *     tags: [catalog]
 *     responses:
 *       "200":
 *         description: OK
 *         content:
 *           application/json:
 *             schema: { type: array, items: { $ref: "#/components/schemas/Category" } }
 *       "500":
 *         description: Internal error
 *         content:
 *           application/json:
 *             schema: { $ref: "#/components/schemas/Problem" }
 */
router.get("/categories", async (_req, res) => {
	try {
		res.json(wire(await catalog.categories()));
	} catch (error) {
		fail(res, error);
	}
});

/**
 * @openapi
 * /brands:
 *   get:
 *     summary: List brands
 *     tags: [catalog]
 *     responses:
 *       "200":
 *         description: OK
 *         content:
 *           application/json:
 *             schema: { type: array, items: { $ref: "#/components/schemas/Brand" } }
 *       "500":
 *         description: Internal error
 *         content:
 *           application/json:
 *             schema: { $ref: "#/components/schemas/Problem" }
 */
router.get("/brands", async (_req, res) => {
	try {
		res.json(wire(await catalog.brands()));
	} catch (error) {
		fail(res, error);
	}
});

/**
 * @openapi
 * /shipping-methods:
 *   get:
 *     summary: List shipping methods
 *     tags: [catalog]
 *     responses:
 *       "200":
 *         description: OK
 *         content:
 *           application/json:
 *             schema:
 *               { type: array, items: { $ref: "#/components/schemas/ShippingMethod" } }
 *       "500":
 *         description: Internal error
 *         content:
 *           application/json:
 *             schema: { $ref: "#/components/schemas/Problem" }
 */
router.get("/shipping-methods", async (_req, res) => {
	try {
		res.json(wire(await catalog.shippingMethods()));
	} catch (error) {
		fail(res, error);
	}
});

/**
 * @openapi
 * /payment-methods:
 *   get:
 *     summary: List payment methods
 *     tags: [catalog]
 *     responses:
 *       "200":
 *         description: OK
 *         content:
 *           application/json:
 *             schema:
 *               { type: array, items: { $ref: "#/components/schemas/PaymentMethod" } }
 *       "500":
 *         description: Internal error
 *         content:
 *           application/json:
 *             schema: { $ref: "#/components/schemas/Problem" }
 */
router.get("/payment-methods", async (_req, res) => {
	try {
		res.json(wire(await catalog.paymentMethods()));
	} catch (error) {
		fail(res, error);
	}
});
